lib/topics_filter.rb

```diff
@@ -92,110 +77,207 @@ def filter_status(status:, category_id: nil)
 
   private
 
-  def filter_created_by_user(usernames:)
-    register_scope(
-      key: :created_by_user,
-      params: {
-        usernames: usernames.map(&:downcase),
-      },
-    ) do |usernames_lower|
-      @scope.joins(:user).where("users.username_lower IN (?)", usernames_lower)
+  def filter_categories(values:)
+    exclude_subcategories_category_slugs = []
+    include_subcategories_category_slugs = []
+
+    values.each do |key_prefix, value|
+      break if key_prefix && key_prefix != "="
+
+      value
+        .scan(
+          /^(?<category_slugs>([a-zA-Z0-9\-:]+)(?<delimiter>[,])?([a-zA-Z0-9\-:]+)?(\k<delimiter>[a-zA-Z0-9\-:]+)*)$/,
+        )
+        .each do |category_slugs, delimiter|
+          (
+            if key_prefix.presence
+              exclude_subcategories_category_slugs
+            else
+              include_subcategories_category_slugs
+            end
+          ).concat(category_slugs.split(delimiter))
+        end
     end
-  end
 
-  def filter_state(state:)
-    case state
-    when "pinned"
-      @scope.where(
-        "topics.pinned_at IS NOT NULL AND topics.pinned_until > topics.pinned_at AND ? < topics.pinned_until",
-        Time.zone.now,
-      )
-    when "bookmarked"
-      return @scope.none unless @guardian.user
+    category_ids = []
 
-      @scope.joins(:topic_users).where(
-        "topic_users.bookmarked AND topic_users.user_id = ?",
-        @guardian.user.id,
+    if exclude_subcategories_category_slugs.present?
+      category_ids =
+        get_category_ids_from_slugs(
+          exclude_subcategories_category_slugs,
+          exclude_subcategories: true,
+        )
+    end
+
+    if include_subcategories_category_slugs.present?
+      category_ids.concat(
+        get_category_ids_from_slugs(
+          include_subcategories_category_slugs,
+          exclude_subcategories: false,
+        ),
       )
-    else
-      return @scope.none if !@guardian.user
+    end
 
-      topic_notification_levels = Set.new
+    if category_ids.present?
+      @scope = @scope.where("topics.category_id IN (?)", category_ids)
+    elsif category_ids.length <
+          (
+            exclude_subcategories_category_slugs.length +
+              include_subcategories_category_slugs.length
+          )
```

Comment:
```suggestion
    elsif (
            exclude_subcategories_category_slugs.length +
              include_subcategories_category_slugs.length
          ) > 0
```

It may need syntax tree formatting, but I think this reads a bit better? `category_ids.length` has to be 0 for us to hit the `elsif` path.

Reply:
Good catch :+1: